### PR TITLE
[Core] Simplify the code to read env variables in RayConfig

### PR DESCRIPTION
**Pull request by kfstorm**
## Why are these changes needed?

Follow up of #16240.

NOTE that this PR requires uppercase env variables. e.g. `RAY_MAX_FUSED_OBJECT_COUNT` rather than `RAY_max_fused_object_count`.

This PR also fixes some bugs:
* Bug: Empty strings are accepted when parsing value.
* Bug: Invalid numbers are accepted. e.g. "123abc" is parsed as "123".

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/common/ray_config.h [hunk 1]
```diff
@@ -23,27 +23,23 @@
...
 /// \param default_value Default value of the config item.
-#define RAY_CONFIG(type, name, default_value)                      \
- private:                                                          \
-  type name##_ = env_##type("RAY_" TOSTRING(name), default_value); \
-                                                                   \
- public:                                                           \
+#define RAY_CONFIG(type, name, default_value)                                     \
+ private:                                                                         \
+  type name##_ =                                                                  \
+      ReadEnv<type>("RAY_" + absl::AsciiStrToUpper(#name), #type, default_value); \
```
Comment (ericl) [inline src/ray/common/ray_config.h:40]: Is it possible to keep the lower-case names? I think it emphasizes:
1. It's an internal system config
2. It's consistent with how gflags env vars work: https://gflags.github.io/gflags/

I'd like to reserve ALL_UPPER_CASE env vars for officially supported configs.

**Comment by ericl** [comment]
Just one request

**Review by rkooo567 — APPROVED** [review]
LGTM. Btw, are there any tests for this env variable system config?

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
Build failure

**Comment by kfstorm** [comment]
@ericl Could you help? I can't solve the compilation failure on Linux. It works on macOS.

**Comment by ericl** [comment]
It seems "ReadEnv" isn't defined in Linux?

```


Use --sandbox_debug to see verbose messages from the sandbox

[0m[91mIn file included from bazel-out/k8-opt/bin/_virtual_includes/ray_common/ray/common/asio/instrumented_io_context.h:21,

                 from bazel-out/k8-opt/bin/_virtual_includes/ray_common/ray/common/asio/periodical_runner.h:19,

                 from src/ray/common/asio/periodical_runner.cc:15:

bazel-out/k8-opt/bin/_virtual_includes/ray_common/ray/common/ray_config.h:103:13: error: explicit specialization in non-namespace scope â€˜class RayConfigâ€™

  103 |   template <>

      |             ^

bazel-out/k8-opt/bin/_virtual_includes/ray_common/ray/common/ray_config.h:105:65: error: template-id â€˜ConvertValue<std::string>â€™ in declaration of primary template

  105 |                                         const std::string &value) {

      |                                                                 ^

bazel-out/k8-opt/bin/_virtual_includes/ray_common/ray/common/ray_config.h:109:13: error: explicit specialization in non-namespace scope â€˜class RayConfigâ€™

  109 |   template <>

      |             ^

bazel-out/k8-opt/bin/_virtual_includes/ray_common/ray/common/ray_config.h:110:83: error: template-id â€˜ConvertValue<bool>â€™ in declaration of primary template

  110 |   bool ConvertValue<bool>(const std::string &type_string, const std::string &value) {

      |                                                                                   ^

bazel-out/k8-opt/bin/_virtual_includes/ray_common/ray/common/ray_config.h:110:8: error: â€˜bool RayConfig::ConvertValue(const string&, const string&)â€™ cannot be overloaded with â€˜std::string RayConfig::ConvertValue(const string&, const string&)â€™

  110 |   bool ConvertValue<bool>(const std::string &type_string, const std::string &value) {

      |        ^~~~~~~~~~~~~~~~~~

bazel-out/k8-opt/bin/_virtual_includes/ray_common/ray/common/ray_config.h:104:15: note: previous declaration â€˜std::string RayConfig::ConvertValue(const string&, const string&)â€™

  104 |   std::string ConvertValue<std::string>(const std::string &type_string,

      |               ^~~~~~~~~~~~~~~~~~~~~~~~~

bazel-out/k8-opt/bin/_virtual_includes/ray_common/ray/common/ray_config.h:116:5: error: too many template-parameter-lists

  116 |   T ReadEnv(const std::string &name, const std::string &type_string, T default_value) {

      |     ^~~~~~~

In file included from bazel-out/k8-opt/bin/_virtual_includes/ray_common/ray/common/asio/instrumented_io_context.h:21,

                 from bazel-out/k8-opt/bin/_virtual_includes/ray_common/ray/common/asio/periodical_runner.h:19,

                 from src/ray/common/asio/periodical_runner.cc:15:

bazel-out/k8-opt/bin/_virtual_includes/ray_common/ray/common/ray_config.h:40:7: error: â€˜ReadEnvâ€™ was not declared in this scope; did you mean â€˜readvâ€™?

   40 |       ReadEnv<type>("RAY_" #name, #type, default_value); \

      |       ^~~~~~~

bazel-out/k8-opt/bin/_virtual_includes/ray_common/ray/common/ray_config.h:40:7: note: in definition of macro â€˜RAY_CONFIGâ€™

   40 |       ReadEnv<type>("RAY_" #name, #type, default_value); \

      |       ^~~~~~~

bazel-out/k8-opt/bin/_virtual_includes/ray_common/ray/common/ray_config.h:40:19: error: expected primary-expression before â€˜>â€™ token

   40 |       ReadEnv<type>("RAY_" #name, #type, default_value); \

      |                   ^

bazel-out/k8-opt/bin/_virtual_includes/ray_common/ray/common/ray_config_def.h:22:1: note: in expansion of macro â€˜RAY_CONFIGâ€™

   22 | RAY_CONFIG(uint64_t, debug_dump_period_milliseconds, 10000)

      | ^~~~~~~~~~

bazel-out/k8-opt/bin/_virtual_includes/ray_common/ray/common/ray_config.h:40:21: error: left operand of comma operator has no effect [-Werror=unused-value]

   40 |       ReadEnv<type>("RAY_" #name, #type, default_value); \

bazel-out/k8-opt/bin/_virtual_includes/ray_common/ray/common/ray_config.h:40:21: note: in definition of macro â€˜RAY_CONFIGâ€™

   40 |       ReadEnv<type>("RAY_" #name, #type, default_value); \

      |                     ^~~~~~

bazel-out/k8-opt/bin/_virtual_includes/ray_common/ray/common/ray_config_def.h:22:54: error: right operand of comma operator has no effect [-Werror=unused-value]

   22 | RAY_CONFIG(uint64_t, debug_dump_period_milliseconds, 10000)

      |                                                      ^~~~~

bazel-out/k8-opt/bin/_virtual_includes/ray_common/ray/common/ray_config.h:40:42: note: in definition of macro â€˜RAY_CONFIGâ€™

   40 |       ReadEnv<type>("RAY_" #name, #type, default_value); \

      |                                          ^~~~~~~~~~~~~

bazel-out/k8-opt/bin/_virtual_includes/ray_common/ray/common/ray_config.h:40:7: error: â€˜ReadEnvâ€™ was not declared in this scope; did you mean â€˜readvâ€™?

   40 |       ReadEnv<type>("RAY_" #name, #type, default_value); \
```

cc @scv119 @iycheng might have better ideas

**Comment by iycheng** [comment]
@kfstorm can you patch this? https://gist.github.com/iycheng/05747954d28ffc2a3404932ef62d0129

The error is because you do specialization in class directly.

**Pull request merged**